crates/mun_gc/src/lib.rs

```diff
@@ -0,0 +1,92 @@
+mod mark_sweep;
+mod ptr;
+mod root_ptr;
+
+pub use mark_sweep::MarkSweep;
+pub use ptr::{GcPtr, HasIndirectionPtr, RawGcPtr};
+pub use root_ptr::GcRootPtr;
+use std::marker::PhantomData;
+
+/// Contains stats about the current state of a GC implementation
+#[derive(Debug, Clone, Default)]
+pub struct Stats {
+    pub allocated_memory: usize,
+}
+
+/// A trait used by the `GcRuntime` to identify an object type.
+pub trait Type: Send + Sync {
+    type Trace: Iterator<Item = GcPtr>;
+
+    /// Returns the size of an object of this type (in bytes).
+    fn size(&self) -> usize;
+
+    /// Returns the alignment of a type (in bytes).
+    fn alignment(&self) -> usize;
+
+    /// Returns an iterator to iterate over all GC objects that are referenced by the given object.
+    fn trace(&self, obj: GcPtr) -> Self::Trace;
+}
+
+/// An object that can be used to allocate and collect memory.
+pub trait GcRuntime<T: Type>: Send + Sync {
+    /// Allocates an object of the given type returning a GcPtr
+    fn alloc(&self, ty: T) -> GcPtr;
+
+    /// Returns the type of the specified `obj`.
+    fn ptr_type(&self, obj: GcPtr) -> T;
+
+    /// Roots the specified `obj`, which keeps its and objects it references alive. Objects marked
+    /// as root, must call `unroot` before they can be collected. An object can be rooted multiple
+    /// times, but you must make sure to call `unroot` an equal number of times before the object
+    /// can be collected.
+    fn root(&self, obj: GcPtr);
+
+    /// Unroots the specified `obj`, potentially allowing it and objects it references to be
+    /// collected. Internally this decrements a root refcount. An object can be rooted multiple
+    /// times, but you must make sure to call `unroot` an equal number of times before the object
+    /// can be collected.
+    fn unroot(&self, obj: GcPtr);
+
+    /// Returns stats about the current state of the runtime.
+    fn stats(&self) -> Stats;
+}
+
+/// An `Event` is an event that can be emitted by a `GcRuntime` through the use of an `Observer`.
+/// This enables tracking of the runtimes behavior which is useful for testing.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Event {
+    /// The GC performed an allocation
+    Allocation(GcPtr),
+
+    /// A GC cycle started
+    Start,
+
+    /// A deallocation took place
+    Deallocation(GcPtr),
+
+    /// A GC cycle ended
+    End,
+}
+
+/// The `Observer` trait allows receiving of `Event`s. A `GcRuntime` can be typed by a `Observer`
+/// which enables optional tracing of events.
```

Comment:
```suggestion
/// The `Observer` trait allows receiving of `Event`s.
```
I think the latter part of the documentation belongs to the `GcRuntime`, since `GcRuntime` depends on `Observer`.